### DISCUSSION TITLE
Get `sqlalchemy.exc.StatementError: (sqlalchemy.exc.MissingGreenlet)` at `append` to a many-to-many (secondary) `relationship` property

### DISCUSSION THREAD

**Discussion by rinarakaki** [sqlalchemy/sqlalchemy#10550] [Usage Questions]
`models.py`

```py
@as_declarative()
class Base:
    pass

class LeftRight(Base):
    __tablename__ = "left_rights"
    left_id: Mapped[int] = mapped_column(ForeignKey("lefts.id"), nullable=False)
    left: Mapped["Left"] = relationship("Left", uselist=False)
    right_id: Mapped[int] = mapped_column(ForeignKey("rights.id"), nullable=False)
    right: Mapped["Right"] = relationship("Right", uselist=False)

class Left(Base):
    rights: Mapped[List["Right"]] = relationship("Right", uselist=True, secondary="left_rights")

class Right(Base):
    lefts: Mapped[List["Left"]] = relationship("Left", uselist=True, secondary="left_rights")
```

`tests.py`

```py
async def test_left_right(session, left_factory, right_factory):
    left = await left_factory()
    right = await right_factory()
    left.rights.append(right)  # ERROR
    session.add(left)
    await session.flush()
    assert True
```

output:

```
E           sqlalchemy.exc.StatementError: (sqlalchemy.exc.MissingGreenlet) greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place?
```

Could someone please help me solve this? The searchability of the solution is always so low in this area. Thank you.

**Comment by zzzeek** [comment]
> The searchability of the solution is always so low in this area. 

so I'm not sure what SQLAlchemy version this is, but when this error is thrown in modern versions it looks like this:

    sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/20/xd2s)

the link there is a real link, which redirects to here:

https://docs.sqlalchemy.org/en/20/errors.html#error-xd2s   which then links to https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#asyncio-orm-avoid-lazyloads - that's the authorititative doc for this error and it should include every possible cause for this issue.

**Reply by rinarakaki** [reply]
The required changes were:

`models.py`

```py
@as_declarative()
class Base(AsyncAttrs):  # inherit AsyncAttrs
    pass
```

`tests.py`

```py
    ...
    (await left.awaitable_attrs.rights).append(right)  # insert await ... .awaitable_attrs
    ...
```

**Reply by rinarakaki** [reply]
1. Is there a way to make the `rights` prop awaitable without intermediate `awaitable_attrs`?
2. Why would it need to SELECT rights first where I only need to INSERT a new row?

**Reply by CaselIT** [reply]
Both questions are answered by the link provided by mike:
1. You can use an eager loader option https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#relationship-loading-techniques
2. That's how a relationship works. It doesn't know that you are only adding an item whe you access it. You can change its behaviour by using https://docs.sqlalchemy.org/en/20/orm/large_collections.html#write-only-relationship

**Reply by rinarakaki** [reply]
No, it doesn't work:

`models.py`

```py
class Left(Base):
    # added lazy="selectin" option
    rights: Mapped[List["Right"]] = relationship("Right", uselist=True, secondary="left_rights", lazy="selectin")
```

`tests.py`

```py
    left.rights.append(right)  # ERROR
```

The link lacks the cases for 1) many-to-many, 2) .append/.remove API and 3) async mode.

**Reply by CaselIT** [reply]
> No, it doesn't work:

what doesn't work? without an example it's quite hard to figure out what you are trying to do

**Reply by rinarakaki** [reply]
The problem settings are as described on the top comment. From there, I added `lazy="selectin"` or `lazy="joined"` option to the `relationship` which is then `Left.rights` attributes because you suggested I could use an eager loader option. But it didn't work: at the line of `left.rights.append(right)`, I keep getting the same error as described on the top comment:

`E           sqlalchemy.exc.StatementError: (sqlalchemy.exc.MissingGreenlet) greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place?`

**Reply by CaselIT** [reply]
You aren't showing how you load an instance. That's the example I'm saying it's missing

**Reply by CaselIT** [reply]
> You aren't showing how you load an instance. That's the example I'm saying it's missing

This complete example works without issues (outside warning because of missing back-populates / viewonly)
```py
import asyncio
from typing import List
from sqlalchemy.orm import as_declarative, Mapped, mapped_column, relationship
from sqlalchemy import ForeignKey
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession


@as_declarative()
class Base:
    pass


class LeftRight(Base):
    __tablename__ = "left_rights"
    left_id: Mapped[int] = mapped_column(ForeignKey("lefts.id"), nullable=False, primary_key=True)
    left: Mapped["Left"] = relationship("Left", uselist=False)
    right_id: Mapped[int] = mapped_column(ForeignKey("rights.id"), nullable=False, primary_key=True)
    right: Mapped["Right"] = relationship("Right", uselist=False)


class Left(Base):
    __tablename__ = "lefts"
    id: Mapped[int] = mapped_column(primary_key=True)
    rights: Mapped[List["Right"]] = relationship(
        "Right", uselist=True, secondary="left_rights", lazy="selectin"
    )


class Right(Base):
    __tablename__ = "rights"
    id: Mapped[int] = mapped_column(primary_key=True)
    lefts: Mapped[List["Left"]] = relationship("Left", uselist=True, secondary="left_rights")


e = create_async_engine("sqlite+aiosqlite://")


async def go():
    async with e.begin() as c:
        await c.run_sync(Base.metadata.create_all)

    async with AsyncSession(e) as s:
        s.add(Left())
        s.add(Right())
        await s.commit()
    async with AsyncSession(e) as s:
        l = await s.get(Left, 1)
        r = await s.get(Right, 1)
        l.rights.append(r)
        await s.commit()

    async with AsyncSession(e) as s:
        l = await s.get(Left, 1)
        print(l.rights)


asyncio.run(go())

```
